net/xia/ppal_ether/main.c

```diff
@@ -191,10 +448,63 @@ static struct ether_interface *eint_init(struct net_device *dev)
 	return eint;
 }
 
+/* Caller must hold RTNL lock, and makes sure that nobody adds entries
+ * in eint->list_interface_common_addr while it's running.
+ */
+static void free_neighs_by_interface(struct ether_interface *eint)
+{
+	struct net_device *dev;
+	struct net *net;
+	struct xip_ppal_ctx *ctx;
+	struct fib_xid_table *xtbl;
+
+	ASSERT_RTNL();
+
+	dev = eint->dev;
+	net = dev_net(dev);
+	ctx = xip_find_my_ppal_ctx_vxt(net, ether_vxt);
+	xtbl = ctx->xpc_xtbl;
+
+	while (1) {
+		struct fib_xid_ether_main *mether;
+		u8 xid[XIA_XID_MAX];
+		struct fib_xid *fxid;
+		u32 bucket;
+
+		/* Obtain xid of the first entry in
+		 * @eint->list_interface_common_addr.
+		 * We use rcu_read_lock() here to allow one to remove
+		 * entries in parallel.
+		 */
+		rcu_read_lock();
+		mether = list_first_or_null_rcu(&eint->
+						list_interface_common_addr,
+					    struct fib_xid_ether_main,
+					    interface_common_addr);
```

Comment:
Instead of removing one mether at a time, you should use `list_for_each_entry_safe()` and move the code below that removes each entry inside of the body of `list_for_each_entry_safe()`.

The reason for that is that, after making many changes to this patch through the code reviews, the current code has a busy wait condition: we get `xid`, get out of the RCU read lock, remove the `fxid` of the table, and call `fxid_free()`. Notice that `fxid` is not going to be released for a while because of RCU; during this period we keep getting the same `xid` from `list_first_or_null_rcu()` and not finding the `xid` in the `xtbl` table. Eventually, RCU releases `fxid` and the code finally moves forward.